pmlc/dialect/comp/ir/ops.td

```diff
@@ -0,0 +1,238 @@
+                                 DeclareOpInterfaceMethods<COMP_ExecEnvOpInterface>,
+                                 DeclareOpInterfaceMethods<COMP_ScheduleOpInterface>]>
+                  , Arguments<(ins COMP_AnyExecEnv:$execEnv,
+                                   Variadic<COMP_AnyEvent>:$depEvents)>
+                  , Results<(outs COMP_AnyEvent:$outEvent)> {
+  let summary = "Schedules function in region for execution on device.";
+  let description = [{
+    This is a wrapper around gpu.launch_func that schedules function
+    to execute on environment and provides synchronization with
+    event dependencies and output event.
+    The region is expected to have only this single gpu.launch_func operation
+    and implicit terminator - comp.schedule_end.
+    Example:
+    %ev = "comp.launch_func"(%execEnv, %depEvent) ( {
+      "gpu.launch_func"(...)
+      "comp.schedule_end"
+    })
+    comp.wait %ev
+  }];
+
+  let regions = (region AnyRegion:$body);
+
+  let verifier = [{ return ::verify(*this); }];
+}
+
+
+def COMP_Submit : COMP_Op<"submit",
+                          [DeclareOpInterfaceMethods<COMP_ExecEnvOpInterface>]>
+                , Arguments<(ins COMP_AnyExecEnv:$execEnv)> {
+  let summary = "Submits all scheduled tasks.";
+
+  let description = [{
+    Forces all previously scheduled tasks to start executing.
+  }];
+
+  let assemblyFormat = [{
+    $execEnv attr-dict `:` type($execEnv)
+  }];
+}
+
+def COMP_Alloc : COMP_Op<"alloc", [MemoryEffects<[MemAlloc<DefaultResource>]>,
+                                   COMP_ExecEnvSupportsMemTrait<"execEnv", "deviceMem">,
+                                   DeclareOpInterfaceMethods<COMP_ExecEnvOpInterface>]>
+               , Arguments<(ins COMP_AnyExecEnv:$execEnv,
+                                Optional<AnyMemRef>:$hostMem)>
+               , Results<(outs AnyMemRef:$deviceMem)> {
+  let summary = "Allocates memory for use in execution environment.";
+
+  let description = [{
+    Optionally accepts host memory, in which case there is guarantee that
+    after allocation new device memory will contain the same data.
+  }];
+
+  let assemblyFormat = [{
+     $execEnv ($hostMem^)? attr-dict `:` functional-type(operands, results)
+  }];
+
+  let verifier = [{ return ::verify(*this); }];
+}
+
+def COMP_Dealloc : COMP_Op<"dealloc", [MemoryEffects<[MemFree<DefaultResource>]>,
+                                       COMP_ExecEnvSupportsMemTrait<"execEnv", "deviceMem">,
+                                       DeclareOpInterfaceMethods<COMP_ExecEnvOpInterface>]>
+                  , Arguments<(ins COMP_AnyExecEnv:$execEnv,
+                                   AnyMemRef:$deviceMem)> {
+  let summary = "Deallocates memory previously allocated for use in executon environment.";
+
+  let assemblyFormat = [{
+    $execEnv $deviceMem attr-dict `:` functional-type(operands, results)
+  }];
+}
+
+def COMP_ScheduleWrite : COMP_Op<"schedule_write",
```

Comment:
How do you want to handle shared-memory devices, and devices that can expose their memory in the system physical address space (for write-combining transfers and such)?

Reply:
In my original draft I have implemented this as `schedule_map : (execenv, memref, event...) -> (event, memref)` and matching `schedule_unmap`, but I have seen no practical use for this so far.
I'd like it to be implemented later, when it would be needed.